src/com/amazon/ion/impl/lite/IonValueLite.java

```diff
@@ -406,34 +402,131 @@ public SymbolTable getSymbolTable()
+            // If the field name's text is unknown, use its sid instead
+            String text = value._fieldName;
+
+            int nameHashCode = text == null
+                ? value._fieldId * sidHashSalt
+                : text.hashCode() * textHashSalt;
+
+            // mixing to account for small text and sid deltas
+            nameHashCode ^= (nameHashCode << 17) ^ (nameHashCode >> 15);
+
+            int fieldHashCode = parent.hashSignature();
+            fieldHashCode = valueHashSalt * fieldHashCode + partial;
+            fieldHashCode = nameHashSalt  * fieldHashCode + nameHashCode;
+
+            // another mix step for each Field of the struct
+            fieldHashCode ^= (fieldHashCode << 19) ^ (fieldHashCode >> 13);
+            return fieldHashCode;
+        }
+
+        void update(int partial, IonValueLite value) {
+            if (parent == null) {
+                valueHash = partial;
+            } else {
+                if (parent instanceof IonStructLite) {
+                    // Additive hash is used to ensure insensitivity to order of
+                    // fields, and will not lose data on value hash codes
+                    valueHash += hashStructField(partial, value);
+                } else {
+                    valueHash = valueHashSalt * valueHash + partial;
+                    // mixing at each step to make the hash code order-dependent
+                    valueHash ^= (valueHash << 29) ^ (valueHash >> 3);
+                }
+            }
+        }
+    }
 
     /**
-     * Since {@link #equals(Object)} is overridden, each concrete class must provide
-     * an implementation of {@link Object#hashCode()}.
+     * Since {@link #equals(Object)} is overridden, each concrete class representing a scalar value must provide
+     * an implementation of {@link #scalarHashCode()}.
      * @return hash code for instance consistent with equals().
      */
-    /*
-     * internally ALL implementations will be delegated too with the SymbolTable
-     * which is to prevent the SymbolTable being continually re-located in complex structures.
-     *
-     * This is near universally true - however does not apply for IonDatagramLite - hence it
-     * explicitly overrides hashCode()
-     */
     @Override
     public int hashCode() {
-        // Supply a lazy symbol table provider, which will call getSymbolTable()
-        // only once it's actually necessary.
-        // This works for all child types with the exception of
-        // IonDatagramLite which has a different, explicit behavior for hashCode()
-        // (hence this method cannot be final).
-        return hashCode(new LazySymbolTableProvider(this));
+        // Branching early for scalars is a significant performance optimization because it avoids allocating an
+        // unnecessary hash stack for these values. Scalars *should* be the most commonly hashed kind of value.
+        if ((_flags & IS_NULL_VALUE) != 0) {
```

Comment:
Using this logic rather than calling `isNullValue()` throughout this change resulted in speedups of up to 7%.

Reply:
This is unexpected, almost incredible. Such trivial, `final` methods ought to get inlined. Bizarre. Okay.

Reply:
The only thing I can think is that it may have been the "straw that broke the camel's back", causing something above it in the call chain to be unable to be inlined.